tests/kernel/futex/src/main.c

```diff
@@ -0,0 +1,379 @@
+		     "k_futex_wake shouldn't wake any thread");
+
+	k_thread_abort(&futex_tid);
+}
+
+void test_futex_wait_forever_wake_from_isr(void)
+{
+	timeout = K_FOREVER;
+
+	k_futex_reset(&simple_futex);
+	atomic_inc(&simple_futex.val);
+
+	k_thread_create(&futex_tid, stack_1, STACK_SIZE,
+			futex_wait_wake_task, &timeout, NULL, NULL,
+			K_PRIO_PREEMPT(0), K_USER | K_INHERIT_PERMS,
+			K_NO_WAIT);
+
+	/* giving time for the futex_wait_wake_task to execute */
+	k_sleep(K_MSEC(100));
+
+	futex_wake_from_isr(&simple_futex);
+
+	/* giving time for the futex_wait_wake_task to execute */
+	k_sleep(K_MSEC(100));
+
+	zassert_true(atomic_get(&simple_futex.val) == 0,
+			"wait forever wake from isr doesn't wake");
+
+	k_thread_abort(&futex_tid);
+}
+
+void test_futex_multiple_threads_wait_wake(void)
+{
+	int woken = 0;
+
+	timeout = K_FOREVER;
+
+	k_futex_reset(&simple_futex);
+
+	for (int i = 0; i < TOTAL_THREADS_WAITING; i++) {
+		atomic_inc(&simple_futex.val);
+		k_thread_create(&multiple_tid[i], multiple_stack[i],
+				STACK_SIZE, futex_wait_wake_task,
+				&timeout, NULL, NULL, K_PRIO_PREEMPT(1),
+				K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+	}
+
+	/* giving time for the other threads to execute  */
+	k_sleep(K_MSEC(500));
+
+	woken = k_futex_wake(&simple_futex, TOTAL_THREADS_WAITING);
+	zassert_true(woken == TOTAL_THREADS_WAITING,
+			"k_futex_wake doesn't wake all threads");
+
+	/* giving time for the other threads to execute  */
+	k_sleep(K_MSEC(500));
+
+	zassert_true(atomic_get(&simple_futex.val) == 0,
+			"wait forever wake doesn't wake all threads");
+
+	for (int i = 0; i < TOTAL_THREADS_WAITING; i++) {
+		k_thread_abort(&multiple_tid[i]);
+	}
+}
+
+void test_multiple_futex_wait_wake(void)
+{
+	int woken = 0;
+
+	timeout = K_FOREVER;
+
+	for (int i = 0; i < TOTAL_THREADS_WAITING; i++) {
+		index[i] = i;
+		k_futex_reset(&multiple_futex[i]);
+		k_thread_create(&multiple_tid[i], multiple_stack[i],
+				STACK_SIZE, futex_multiple_wait_wake_task,
+				&timeout, &index[i], NULL, K_PRIO_PREEMPT(1),
+				K_USER | K_INHERIT_PERMS, K_NO_WAIT);
+	}
+
+	/* giving time for the other threads to execute  */
+	k_sleep(K_MSEC(500));
+
+	for (int i = 0; i < TOTAL_THREADS_WAITING; i++) {
+		woken = k_futex_wake(&multiple_futex[i], 1);
+		zassert_true(woken == 1,
+			"k_futex_wake doesn't wake the thread");
+	}
+
+	/* giving time for the other threads to execute  */
+	k_sleep(K_MSEC(500));
+
+	zassert_true(atomic_get(&simple_futex.val) == 0,
+			"wait forever wake doesn't wake all threads");
+
+	for (int i = 0; i < TOTAL_THREADS_WAITING; i++) {
+		k_thread_abort(&multiple_tid[i]);
+	}
+}
+
+/* ztest main entry*/
+void test_main(void)
+{
+	struct k_mem_partition *parts[] = {&futex_partition,
+		&ztest_mem_partition};
+
+	k_mem_domain_remove_thread(k_current_get());
+	k_mem_domain_init(&futex_domain, ARRAY_SIZE(parts), parts);
+	k_mem_domain_add_thread(&futex_domain, k_current_get());
+
+	ztest_test_suite(test_futex,
+			 ztest_unit_test(test_futex_wait_forever_wake),
```

Comment:
shouldn't all of these except test_futex_wait_forever_wake_from_isr() be running in user mode, with ztest_user_unit_test()?

as it stands right now, I don't k_futex_wake() being called from user mode anywhere

Reply:
yes, the case of k_futex_wake being called from user mode is added